src/transformStateWithClones.js

```diff
@@ -7,29 +7,35 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  const result = [];
-  const copyOfState = { ...state };
+  let currentState = { ...state };
+  const resultStates = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        currentState = {
+          ...currentState, ...action.extraData,
+        };
+        break;
 
-  for (const element of actions) {
-    switch (element.type) {
       case 'removeProperties':
-        for (const item of element.keysToRemove) {
-          delete copyOfState[item];
+        for (const key of action.keysToRemove) {
+          delete currentState[key];
```

Comment:
We want to avoid nested loops, so here you can use the `forEach` construction.

Reply:
@Patrykmclaren14 keep in mind that `forEach` is also a type of loop ;)
Time complexity is the same